test/jdk/com/sun/jdi/InterruptHangTest.java

```diff
@@ -28,93 +28,220 @@
+        } else if (aggressiveMode) {
+            interruptorThread =
+                DebuggeeWrapper.newThread(new AggressiveInterruptor(Thread.currentThread()));
+            interruptorThread.start();
+        } else {
+            interruptorThread = null; // we are in "remote" interrupt mode
         }
 
         // Debugger will keep stepping thru this loop
-        for (int ii = 0; ii < 200; ii++) {
-            answer++;
+        for (int ii = 0; ii < INTERRUPTS_EXPECTED; ii++) {
+            boolean wasInterrupted = false;
             try {
-                // Give other thread a chance to run
+                // Give other thread a chance to interrupt
                 Thread.sleep(100);
             } catch (InterruptedException ee) {
-                System.out.println("Debuggee interruptee: interrupted at iteration: "
-                                   + ii);
+                answer++;
+                wasInterrupted = true;
+                boolean isInterrupted = Thread.currentThread().isInterrupted();
+                System.out.println("Debuggee interruptee(" + ii + "): isInterrupted(" + isInterrupted + ")");
+                if (preciseMode) {
+                    // When Thread.sleep() is interrupted, the interrupted status of the thread
+                    // should remain cleared (since InterruptedException was thrown), and the
+                    // the next interrupt won't come in until after the sync.notify() below.
+                    // Note this is not true for the aggressive and remote modes since
+                    // interrupts can come in at any time, even while we are handling
+                    // an intrrupt.
+                    if (isInterrupted) {
+                        throw new RuntimeException("Thread is interrupted but shouldn't be.");
+                    }
+                    synchronized(InterruptHangTarg.sync) {
+                        // Let the interruptor thread know it can start interrupting again
+                        sync.notify();
+                    }
+                }
+            }
+            // Every Thread.sleep() call should get interrupted
+            if (!wasInterrupted) {
+                throw new RuntimeException("Thread was never interrupted during sleep: " + ii);
+            }
+        }
+
+        if (interruptorThread != null) {
+            synchronized(InterruptHangTarg.sync) {
+                // Kill the interrupter thread
+                interruptorThread.interrupt();
             }
         }
-        // Kill the interrupter thread
-        interruptorThread.interrupt();
+
+        if (!remoteMode) { // we don't count interruptsSent when in remote mode
+            System.out.println("interrupts sent: " + interruptsSent);
+        }
+
+        // When in precise mode, make sure that every interrupt sent resulted in
+        // an InterruptedException. Note the interruptor always ends up sending
+        // one extra interrupt at the end.
```

Comment:
The extra interrupt is because the interruptor will get another shot at interrupting once the interuptee does the following on the last interation (not this is done on every iteration of the main interruptee loop)

                    synchronized(InterruptHangTarg.sync) {
                        // Let the interruptor thread know it can start interrupting again
                        sync.notify();
                    }

Avoiding this last notify() would require that it be conditionally avoided on the last iteration of the loop (by comparing ii to INTERRUPTS_EXPECTED), and then as a result set some sort of flag so the PreciseInterruptor doesn't do another Thread.interrupt() after the interruptee wakes it up with the sync.notify(). 

Having a check in the middle of a loop to see if we are on the last iteration isn't exactly pretty either. I felt having the extra interrupt() was a better choice.

Reply:
Okay. It makes sense.